core/dbt/include/global_project/macros/materializations/incremental/incremental.sql

```diff
@@ -28,13 +32,23 @@
       {% set build_sql = create_table_as(False, intermediate_relation, sql) %}
       {% set need_swap = true %}
       {% do to_drop.append(backup_relation) %}
+  
   {% else %}
-      {% set tmp_relation = make_temp_relation(target_relation) %}
       {% do run_query(create_table_as(True, tmp_relation, sql)) %}
+      {% set schema_changed = check_for_schema_changes(tmp_relation, target_relation) %}
+  
+      {% if schema_changed and on_schema_change == 'full_refresh' %}
+        {# placeholder for full refresh from permanent table #}
+      {% endif %}
```

Comment:
Okay, I've thought about it some more, and I think we have two options in the case of `on_schema_change == 'full_refresh'`:
1. Create the temp table first, compare the schemas. If the schemas mismatch, proceed down a copy of the full-refresh conditional path: abandon the temp table, create the _full_ table, proceed with swap/atomic replacement as appropriate. (Downside: we'll need to reimplement the logic.) If the schemas match, proceed down incremental conditional branch.
2. Compare the schema _before_ creating the temp or full table. How? I think we could use [`get_columns_in_query`](https://github.com/fishtown-analytics/dbt/blob/85e2c897945e6372fe3657075b2c5ea4b068d3e0/core/dbt/include/global_project/macros/adapters/common.sql#L1-L15), to return just the list of column names. Then we can compare against the current table; if they don't match, proceed down usual full-refresh path; if they do match, proceed down usual incremental path. No need to reimplement the logic.
```sql
{% set schema_change_full_refresh = False %}
{% if on_schema_change == 'full_refresh' %}
  {# could wrap the logic below in a macro / adapt check_for_schema_changes to account for it #}
  {% set source_columns = get_columns_in_query(sql) %}
  {% set target_columns = adapter.get_columns_in_relation(target_relation) %}
  {%- set source_not_in_target = diff_columns(source_columns, target_columns) -%}
  {%- set target_not_in_source = diff_columns(target_columns, source_columns) -%}
  {%- set schema_change_full_refresh = (source_not_in_target or target_not_in_source) %}

{% if existing_relation is none %}
      {% set build_sql = create_table_as(False, target_relation, sql) %}

{% elif should_full_refresh() or existing_relation.is_view or schema_change_full_refresh %}
-- normal full-refresh 

{% else %}
-- normal incremental
```

I'm inclined toward option #2, since it minimizes the duplication of logic, _but_ it would be an acknowledgment that `on_schema_change: full_refresh` will not be able to detect data type mismatches.